### ISSUE TITLE
Codec split

### ISSUE THREAD

**Issue by xplosunn** [softwaremill/tapir#323]
Currently a codec aggregates the concepts of encoding and decoding. This means that if a project only has an encoder or a decoder for a given type, it is not possible to integrate with tapir without implementing the other, which might not make sense in the context of the project. 

Do you think makes sense to split those concerns?

**Comment by ghostbuster91** [comment]
Yes, codecs aggregate both of these concepts because they are used by endpoints which are self contained objects ready for interpretation. The demand for encoding or decoding depends on the position of a particular entity in a given endpoint, but also on the type of an interpreter (e.g. server-inputs will need decoders while server-outputs will need encoders). Right now once we create endpoints they can be interpreted by any kind of an interpreter without any additional work. Of course encoders and decoders have to be in sync, so both the server and the client interpretations will work together. Now this is easily achievable since encode and decode functions are kept in one place.

Demanding only encoders or decoders would meant shifting this requirement to the interpretation point. This can cause related encoders and decoders to diverge. It would also imply that an endpoint is no longer a self contained object, since to interpret it you need to provide relevant codecs. 

To sum up - apart from the technical problems we may face will trying to do that, I am more concerned whether we should do it at all. Anyway this are my concerns, maybe @adamw will have a different opinion.

**Comment by adamw** [comment]
As @ghostbuster91 writes, it's not possible currently. Not only e.g. `stringBody` can be used as an input or output, but an endpoint can be interpreted as a client or server.

Moreover, we can't move codec-capturing to the interpretation point, as we don't know the media types (e.g. do we need a json-codec for `Person`, or a xml-codec).

To do this, we would need to capture more information in the type signature, and that would complicate the types a lot. The tradeoff that tapir took - by design - is to require bi-directional codec always, but have simpler, more readable types.

So sorry to say, but that won't be possible.
